crates/valence/src/client/event.rs

```diff
@@ -852,7 +860,27 @@ fn handle_one_packet(
             });
         }
         C2sPlayPacket::ClickSlotC2s(p) => {
-            if p.slot_idx < 0 {
+            let open_inv = q
+                .open_inventory
+                .as_ref()
+                .and_then(|open| inventories.get(open.entity).ok());
+            if !crate::inventory::validate_click_slot_impossible(&p, &q.inventory, open_inv) {
+                debug!("client {:#?} invalid click slot packet: {:#?}", q.entity, p);
+                return Ok(true);
+            }
+            if !crate::inventory::validate_click_slot_item_duplication(
+                &p,
+                &q.inventory,
+                open_inv,
+                &q.cursor_item,
+            ) {
+                debug!(
+                    "client {:#?} click slot packet tried to incorrectly modify items: {:#?}",
+                    q.entity, p
+                );
+                return Ok(true);
+            }
```

Comment:
If a validation error does occur, the client can go out of sync. We probably need to send a `InventoryS2c` to fix this.